File: 2eq_ssa.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Author: Ruth Kristianingsih <ruth.kristianingsih30@gmail.com>

from math import log, sqrt
#from random import random
import numpy as np

#MAX_TIME = 50
SELECT_FLAG = True

# Given variables
kappa = 1
alpha = 1
gamma = 0.1
sigma = 0.1


#def uniform_dist(min, max):
#	return min + (max - min) * random()

def ssa_gillespie(max_time, R_init, p_init):
	# Initialize populations
	R = R_init
	p = p_init
	# Initialize time
	time = 0.0
	step = 0

	with open("results-ssa.csv", "w") as file:
		# Print initial values to file
		file.write("%f %d %d\n" % (time, R, p))
		# Main loop
		while (time < max_time):
			# Random numbers
			r1 = np.random.uniform(0, 1)
			r2 = np.random.uniform(0, 1)
#			print(r1, r2)

			# Transition rates
			w1 = gamma * R
			w2 = sigma * p
			w3 = alpha * R
			w4 = alpha * R * p
			w5 = kappa * R * p
			w6 = kappa * R * R * p
			w0 = w1 + w2 + w3 + w4 + w5 + w6
			print(R, p, w0)

			# Timestep tau
			tau = 1/w0 * log(1/r1)

			# Compute at time + tau
			if 0 <= r2 < w1/w0:
				R -= 1
			elif w1/w0 <= r2 < (w1+w2)/w0:
				p -= 1
			elif (w1+w2)/w0 <= r2 < (w1+w2+w3)/w0:
				p += 1
			elif (w1+w2+w3)/w0 <= r2 < (w1+w2+w3+w4)/w0:
				p -= 1
			elif (w1+w2+w3+w4)/w0 <= r2 < (w1+w2+w3+w4+w5)/w0:
				R += 1
			elif (w1+w2+w3+w4+w5)/w0 <= r2 < (w1+w2+w3+w4+w5+w6)/w0:
				R -= 1

			# Update time step
			time += tau
			step += 1
			# Print values to file
			file.write("%f %d %d\n" % (time, R, p))
	# Close file
	while not file.closed:
		file.close()

ssa_gillespie(100, 0.5, 0)



#def qssa_gillespie(X1_init):
#	# Initialize populations
#	X1 = X1_init
#	n1 = X1/S
#	P = (n1**2)/( (n1**2) + nu11)
#	X2 = np.random.binomial(E, P)
#	# Initialize time
#	time = 0.0
#	step = 0
#
#	with open("results-qssa.csv", "w") as file:
#		# Print initial values to file
#		file.write("%f %d %d\n" % (time, X1, X2))
#		# Main loop
#		while (time < MAX_TIME):
#			# Compute P and X2
#			P = (n1**2)/( (n1**2) + nu11)
#			X2 = np.random.binomial(E, P)
#
#			# Random numbers
#			z1 = uniform_dist(0,1)
#			z2 = uniform_dist(0,1)
#
#			# Transition rates
#			w1 = E * (E * R + kappa1 * X2 )
#			w2 = E**2 * (kappa2 * n1)
#			w3 = (n1**2) * (E - X2) * E
#			w4 = nu11 * X2 * E
#			w0 = w1 + w2 + w3 + w4
#
#			# Timestep tau
#			tau = 1/w0 * log(1/z1)
#			if tau < 0:
#				print(tau, z1, "It's going back in time.")
#				break
#
#			# Compute at time + tau
#			if 0 <= z2 < w1/w0:
#				n1 += 1
#			elif w1/w0 <= z2 < (w1+w2)/w0:
#				n1 -= 1
#			elif (w1+w2)/w0 <= z2 < (w1+w2+w3)/w0:
#				n1 -= 2
#			elif (w1+w2+w3)/w0 <= z2 <= 1:
#				n1 += 2
#
#			# Update X1
#			X1 = n1 * S
#			# Update time step
#			time += tau
#			step += 1
#			# Print values to file
#			file.write("%f %d %d\n" % (time, X1, X2))
#	# Close file
#	while not file.closed:
#		file.close()


#assumption = input("Which simulation do you want to run? (SSA/QSSA): ")
#
#if assumption.upper() == "SSA" or assumption.upper() == "S":
#	# Stochastic Simulation Algorithm Gillespie
#	print("Simulating Stochastic Simulation Algorithm Gillespie...")
#	# Input initial values
#	if SELECT_FLAG:
#		X1_init = float(input("Select the initial population for X1: "))
#		X2_init = float(input("Select the initial population for X2: "))
#	else:
#		X1_init = 0
#		X2_init = 0
#	# Run the simulation
#	ssa_gillespie(X1_init, X2_init)
#	print("\nThe results can be found on results-ssa.csv")
#
#elif assumption.upper() == "QSSA" or assumption.upper() == "Q":
#	# Quasi-Steady State Assumption Gillespie
#	print("Simulating Quasi-Steady State Assumption Gillespie...")
#	# Input initial values
#	if SELECT_FLAG:
#		X1_init = float(input("Select the initial population for X1: "))
#	else:
#		X1_init = 0
#	# Run the simulation
#	qssa_gillespie(X1_init)
#	print("\nThe results can be found on results-qssa.csv")
#
#else:
#	print("That's not a valid answer")

File: 2eq_ssa.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

float uniform_dist(float, float);
void ssa_gillespie(float[], float *);

// Random Uniform Distribution
float uniform_dist(float min, float max) {
	rand(); // Fix for random
	float scale = rand() / (float) RAND_MAX; /* [0, 1.0] */
	return min + scale * ( max - min );      /* [min, max] */
}

void ssa_gillespie(float X[], float *params) {
	float r1 = uniform_dist(0, 1);
	float r2 = uniform_dist(0, 1);

	// Transition rates
	float *w = calloc(7, sizeof(float));
	float *wsum = calloc(7, sizeof(float));

	w[1] = params[2] * X[1];
	w[2] = params[0] * X[1] * X[1] * X[2] / pow(1000, 2);
	w[3] = params[3] * X[2];
	w[4] = params[1] * X[1] * X[2] / 1000;
	w[5] = params[1] * X[1];
	w[6] = params[0] * X[1] * X[2] / 1000;

	// Total Number of all of the transition rates above
	for (int i = 1; i <= 6; i++) {
		w[0] += w[i];
	}

	// Partial sums of the rates (to be evaluated in the conditionals)
	wsum[0] = 0;

	for (int i = 1; i <= 6; i++) {
		for (int j = 1; j <= i; j++) {
			wsum[i] += w[j];
		}
		wsum[i] /= w[0];
	}

	// Timestep tau
	float tau = 1/w[0] * log(1/r1);

	// printf("%f, %f\n", r2, w0);

	// Compute at time + tau
	if ( (wsum[0] <= r2) & (r2 < wsum[1]) ) {
		X[1] -= 1;
	} else if ( (wsum[1] <= r2) & (r2 < wsum[2]) ) {
		X[1] -= 1;
	} else if ( (wsum[2] <= r2) & (r2 < wsum[3]) ) {
		X[2] -= 1;
	} else if ( (wsum[3] <= r2) & (r2 < wsum[4]) ) {
		X[2] -= 1;
	} else if ( (wsum[4] <= r2) & (r2 < wsum[5]) ) {
		X[2] += 1;
	} else if ( (wsum[5] <= r2) & (r2 < wsum[6]) ) {
		X[1] += 1;
	}
	// Update time step
	X[0] += tau;
}

int main(int argc, char const *argv[]) {
	// Parameters
	float *params = calloc(4, sizeof(float));
	params[0] = atof(argv[4]);  // kappa
	params[1] = atof(argv[5]);  // omega
	params[2] = atof(argv[6]);  // gamma
	params[3] = atof(argv[7]);  // sigma
	srand((unsigned)time(NULL));

	float X[] = { 0, atof(argv[2]), atof(argv[3]) };
	float max_time = atof(argv[1]);

	printf("%f, %f, %f\n", X[0], X[1], X[2]);

	unsigned long int step = 0;

	while (X[0] < max_time) {
		ssa_gillespie(X, params);
		step += 1;
		printf("%f, %f, %f\n", X[0], X[1], X[2]);
	}

	return 0;
}
